Source/WindowsShellExtension/dllmain.cpp

```diff
@@ -0,0 +1,299 @@
+    bool RegGetBool(HKEY hKey, const std::wstring& subKey, const std::wstring& value) {
+        DWORD dataSize{};
+        LONG retCode = RegGetValue(
+            hKey,
+            subKey.c_str(),
+            value.c_str(),
+            RRF_RT_REG_SZ,
+            nullptr,
+            nullptr,
+            &dataSize
+        );
+        if (retCode != ERROR_SUCCESS)
+            throw std::runtime_error("Cannot read string from registry.");
+        std::wstring data;
+        data.resize(dataSize / sizeof(wchar_t));
+        retCode = ::RegGetValue(
+            hKey,
+            subKey.c_str(),
+            value.c_str(),
+            RRF_RT_REG_SZ,
+            nullptr,
+            &data[0],
+            &dataSize
+        );
+        if (retCode != ERROR_SUCCESS)
+            throw std::runtime_error("Cannot read string from registry.");
+        DWORD stringLengthInWchars = dataSize / sizeof(wchar_t);
+        --stringLengthInWchars; // Exclude the NUL written by the Win32 API
+        data.resize(stringLengthInWchars);
+        if (data.compare(L"true") == 0)
+            return true;
+        if (data.compare(L"false") == 0)
+            return false;
+        throw std::runtime_error("Not a boolean.");
+    }
+}
+
+struct ExplorerCommandHandler : public winrt::implements<ExplorerCommandHandler, IExplorerCommand> {
+public:
+    // IExplorerCommand implementation:
+
+    IFACEMETHODIMP GetTitle(_In_opt_ IShellItemArray* items, _Outptr_ PWSTR* name) {
+        // Provide name for display in File Explorer context menu entry
+        UNREFERENCED_PARAMETER(items);
+        return SHStrDupW(menu_entry_title, name);
+    }
+
+    IFACEMETHODIMP GetIcon(_In_opt_ IShellItemArray* items, _Outptr_ PWSTR* icon) {
+        // Provide icon for display in File Explorer context menu entry
+        // Get path to application exe and use it as source for icon
+        UNREFERENCED_PARAMETER(items);
+        std::filesystem::path module_path{ wil::GetModuleFileNameW<std::wstring>(wil::GetModuleInstanceHandle()) };
+        module_path = module_path.remove_filename();
+        module_path /= exe_filename;
+        return SHStrDupW(module_path.c_str(), icon);
+    }
+
+    IFACEMETHODIMP GetToolTip(_In_opt_ IShellItemArray* items, _Outptr_ PWSTR* infoTip) {
+        UNREFERENCED_PARAMETER(items);
+        *infoTip = nullptr;
+        return E_NOTIMPL;
+    }
+
+    IFACEMETHODIMP GetCanonicalName(_Out_ GUID* guidCommandName) {
+        *guidCommandName = GUID_NULL;
+        return S_OK;
+    }
+
+    IFACEMETHODIMP GetState(_In_opt_ IShellItemArray* items, _In_ BOOL okToBeSlow, _Out_ EXPCMDSTATE* cmdState) {
+        // Provide state of File Explorer context menu entry
+        // Hide it if registry setting indicates that it should be disabled, else it is enabled
+        UNREFERENCED_PARAMETER(items);
```

Comment:
My understanding he is that you have the items name so you can check if it is a directory or not, and `ECS_HIDDEN` if there are only directories shell extension for directories is not activated, it does not seem too complicated.

Does it make sense?

Reply:
Probably. Need to make sure it does not take too long or else need check okToBeSlow and I am not sure what will happen if okToBeSlow is false. Maybe you can add it later on.

Reply:
> Need to make sure it does not take too long 

A possibility is to launch the check only if there are few items.
Also, we stop the check when there is at least 1 file.
My guess is that slowness will be very rare, people usually don't select 100 folder, it is more a couple of files + folder so a file will be caught early.

> if okToBeSlow is false

In that case we may just ignore the check.

Reply:
Putting the link of the doc for this here for reference: https://learn.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-iexplorercommand-getstate